server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -695,26 +682,28 @@ UploadedMetadataResults writeMetadataInParallel(
         } catch (InterruptedException ex) {
             exceptionList.forEach(ex::addSuppressed);
             RemoteStateTransferException exception = new RemoteStateTransferException(
-                String.format(
-                    Locale.ROOT,
-                    "Timed out waiting for transfer of metadata to complete - %s",
-                    String.join(", ", uploadTasks.keySet())
-                ),
+                String.format(Locale.ROOT, "Timed out waiting for transfer of metadata to complete - %s", String.join(", ", uploadTasks)),
                 ex
             );
             Thread.currentThread().interrupt();
             throw exception;
         }
         if (!exceptionList.isEmpty()) {
             RemoteStateTransferException exception = new RemoteStateTransferException(
+                String.format(Locale.ROOT, "Exception during transfer of following metadata to Remote - %s", String.join(", ", uploadTasks))
+            );
+            exceptionList.forEach(exception::addSuppressed);
+            throw exception;
+        }
+        if (results.size() != uploadTasks.size()) {
```

Comment:
do we have test covering this case?

Reply:
Yes, added a unit test covering this case.